File: app/src/main/java/com/sofps/inspirationalquotes/util/LanguagePreferences.java
package com.sofps.inspirationalquotes.util;

import android.content.SharedPreferences;
import java.util.Arrays;
import java.util.Locale;

public class LanguagePreferences implements SharedPreferences.OnSharedPreferenceChangeListener {

    public interface LanguagePreferencesListener {
        void onLanguageChange();
    }

    private static final String PREF_LANGUAGE = "language";
    private static final String[] SUPPORTED_LANGUAGES = { "EN", "ES" };
    private static final String DEFAULT_LANGUAGE = "EN";

    private final SharedPreferences mSharedPreferences;
    private LanguagePreferencesListener mLanguagePreferencesListener;

    public LanguagePreferences(SharedPreferences sharedPreferences) {
        mSharedPreferences = sharedPreferences;
        mSharedPreferences.registerOnSharedPreferenceChangeListener(this);
    }

    @Override
    public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) {
        if (mLanguagePreferencesListener != null && PREF_LANGUAGE.equals(key)) {
            mLanguagePreferencesListener.onLanguageChange();
        }
    }

    public String getLanguage() {
        String language = mSharedPreferences.getString(PREF_LANGUAGE, null);
        if (language == null) {
            // The language preference is not set, set it
            language = Locale.getDefault().getLanguage().toUpperCase();
            if (!Arrays.asList(SUPPORTED_LANGUAGES).contains(language)) {
                language = DEFAULT_LANGUAGE;
            }
            setLanguage(language);
        }
        return language;
    }

    public void setLanguage(String language) {
        mSharedPreferences.edit().putString(PREF_LANGUAGE, language).apply();
    }

    public void setLanguagePreferencesListener(LanguagePreferencesListener listener) {
        mLanguagePreferencesListener = listener;
    }

    public void unregisterListeners() {
        mSharedPreferences.unregisterOnSharedPreferenceChangeListener(this);
    }
}

File: app/src/main/java/com/sofps/inspirationalquotes/util/LanguagePreferences.kt
package com.sofps.inspirationalquotes.util

import android.content.SharedPreferences
import java.util.Locale

class LanguagePreferences(
        private val sharedPreferences: SharedPreferences
) : SharedPreferences.OnSharedPreferenceChangeListener {

    companion object {

        private const val PREF_LANGUAGE = "language"
        private const val DEFAULT_LANGUAGE = "EN"
        private val SUPPORTED_LANGUAGES = arrayOf("EN", "ES")
    }

    private var languagePreferencesListener: LanguagePreferencesListener? = null

    // The language preference is not set, set it
    var language: String
        get() {
            var languageAux = sharedPreferences.getString(PREF_LANGUAGE, null)
            if (languageAux == null) {
                languageAux = Locale.getDefault().language.toUpperCase()
                if (!SUPPORTED_LANGUAGES.contains(languageAux)) {
                    languageAux = DEFAULT_LANGUAGE
                }
                language = languageAux
            }
            return languageAux
        }
        set(language) = sharedPreferences.edit().putString(PREF_LANGUAGE, language).apply()

    interface LanguagePreferencesListener {
        fun onLanguageChange()
    }

    init {
        sharedPreferences.registerOnSharedPreferenceChangeListener(this)
    }

    override fun onSharedPreferenceChanged(sharedPreferences: SharedPreferences, key: String) {
        if (languagePreferencesListener != null && PREF_LANGUAGE == key) {
            languagePreferencesListener!!.onLanguageChange()
        }
    }

    fun setLanguagePreferencesListener(listener: LanguagePreferencesListener) {
        languagePreferencesListener = listener
    }

    fun unregisterListeners() {
        sharedPreferences.unregisterOnSharedPreferenceChangeListener(this)
    }
}
